### PR TITLE
bugfix

### PR DESCRIPTION
**Pull request by YaoJiayi**
Fixing bug at https://github.com/LMCache/LMCache/issues/122

Also add support for full prefill and partial prefill in a single batch. However, this needs to be end-to-end tested (@Shaoting-Feng).

### CODE REVIEW

File: lmcache_vllm/vllm_adapter.py [hunk 1]
```diff
@@ -93,12 +93,15 @@ def lmcache_should_retrieve(
     if not has_engine or kv_caches is None:
         return False
 
-    prefill_meta = model_input.attn_metadata.prefill_metadata
-
+    attn_meta = model_input.attn_metadata#.prefill_metadata
```
Comment (KuntaiDu) [inline lmcache_vllm/vllm_adapter.py]: Maybe remove the `#.prefill_metadata` at the end

File: lmcache_vllm/vllm_adapter.py [hunk 2]
```diff
@@ -93,12 +93,15 @@ def lmcache_should_retrieve(
...
+    prefill_meta = attn_meta.prefill_metadata
+    
     # check if the current run is profiling
     is_profile_run = (kv_caches is None) or (kv_caches[0] is None)
     # check if the current run is prefill
-    is_prefill_run = prefill_meta is not None
+    is_prefill_run = ((attn_meta.num_prefills == len(model_input.seq_lens))\
+        and prefill_meta is not None)
+    #prefill_meta is not None
```
Comment (KuntaiDu) [inline lmcache_vllm/vllm_adapter.py]: same here

File: lmcache_vllm/vllm_adapter.py [hunk 3]
```diff
@@ -133,8 +137,11 @@ def lmcache_should_store(
 
     # check if the current run is profiling
     is_profile_run = (kv_caches is None) or (kv_caches[0] is None)
+    
+    # FIXME(Jiayi): need to support multiple prefills in a single batch
     # check if the current run is prefill
-    is_prefill_run = prefill_meta is not None
+    is_prefill_run = ((attn_meta.num_prefills == len(model_input.seq_lens))\
+        and (prefill_meta is not None)) #prefill_meta is not None
```
Comment (KuntaiDu) [inline lmcache_vllm/vllm_adapter.py]: same here

File: lmcache_vllm/vllm_injection.py [hunk 4]
```diff
@@ -118,8 +118,11 @@ def new_execute_model(
  
     if not self.is_driver_worker:
         return []
- 
+
+    # Jiayi: this call back calls `_process_model_outputs`
+    # in vllm/engine/llm_engine.py
     if model_input.async_callback is not None:
+        #logger.debug(f"Model input after running: {model_input}")
```
Comment (KuntaiDu) [inline lmcache_vllm/vllm_injection.py]: same here

**Review by KuntaiDu — APPROVED** [review]
LGTM!

**Comment by Shaoting-Feng** [comment]
Offline E2E Tests:

| Request Number | 1  | 2            | 3            | 4                                   | 5            | 6            | 7                                   | 8            | 9            | 10                                  |
|----------------|----|--------------|--------------|-------------------------------------|--------------|--------------|-------------------------------------|--------------|--------------|-------------------------------------|
| Request Type   | First Request | Partial Prefill  | Full Prefill | Full & Partial Prefill in a Single Batch | Partial Prefill  | Full Prefill | Full & Partial Prefill in a Single Batch | Partial Prefill  | Full Prefill | Full & Partial Prefill in a Single Batch |
| TTFT (s) | 0.176 | 0.047 | 0.098 | 0.079 | 0.054 | 0.102 | 0.075 | 0.041 | 0.094 | 0.076 |

**Pull request merged**